test/graph_test.cpp

```diff
@@ -66,6 +67,42 @@ TEST_P(GraphTest, IFconnectedComponentsAlgRunTHENupdateLocked) {
   ASSERT_THROW(g.update({{1,2}, DELETE}), UpdateLockedException);
 }
 
+TEST_P(GraphTest, TestSupernodeRestoreAfterCCFailure) {
+  write_configuration(false, GetParam());
+  const std::string fname = __FILE__;
+  size_t pos = fname.find_last_of("\\/");
+  const std::string curr_dir = (std::string::npos == pos) ? "" : fname.substr(0, pos);
+  std::ifstream in{curr_dir + "/res/multiples_graph_1024.txt"};
+  node_id_t num_nodes;
+  in >> num_nodes;
+  edge_id_t m;
+  in >> m;
+  node_id_t a, b;
+  Graph g{num_nodes};
+  while (m--) {
+    in >> a >> b;
+    g.update({{a, b}, INSERT});
+  }
+  g.set_verifier(std::make_unique<FileGraphVerifier>(curr_dir + "/res/multiples_graph_1024.txt"));
+  g.should_fail_CC();
+
+  // flush to make sure copy supernodes is consistent with graph supernodes
+  g.bf->force_flush();
+  GraphWorker::pause_workers();
```

Comment:
let it be known I spent the entire day debugging the graph functions when this was the real reason the test was failing.

Reply:
Oh no :D

Reply:
Oh no :D